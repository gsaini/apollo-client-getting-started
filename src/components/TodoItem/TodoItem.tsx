import React from 'react';
import { ToDo } from '../../types';

type TodoItemProps = {
    todo: ToDo;
    onChange: (todo: Pick<ToDo, 'id' | 'completed'>) => void
};

export const TodoItem = ({ todo, onChange }: TodoItemProps): JSX.Element => {
    const { id, title, completed } = todo;

    return (
        <>
            <div className="Todo__item">
                <label htmlFor={`${id}-checkbox`}>{title}</label>
                <input
                    id={`${id}-checkbox`}
                    type="checkbox"
                    checked={completed || false}
                    onChange={() => onChange({ id, completed: !completed })}
                />
            </div>
        </>);
}