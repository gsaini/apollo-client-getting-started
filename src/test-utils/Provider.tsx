import React from "react";
import { act, render, RenderResult } from "@testing-library/react";
import { MockedProvider, MockedProviderProps } from '@apollo/client/testing';
import { cache } from '../apollo';


const customRender = async (
    childComponent: JSX.Element,
    options?: MockedProviderProps
): Promise<RenderResult> => {
    const { mocks = [], resolvers = {} } = options || {};
    const wrapper = render(<MockedProvider
        mocks={mocks}
        resolvers={resolvers}
        cache={cache}>
        {childComponent}
    </MockedProvider>)

    await act(() => new Promise((resolve) => setTimeout(resolve, 0)));

    return wrapper;
}

export { customRender as render };