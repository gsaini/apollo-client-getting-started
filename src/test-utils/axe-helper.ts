import { configureAxe } from "jest-axe";

// https://github.com/dequelabs/axe-core/blob/develop/doc/rule-descriptions.md

export const axe = configureAxe({
  rules: {
    // for demonstration only, don't disable rules that need fixing.
    "image-alt": { enabled: false },
  },
});
