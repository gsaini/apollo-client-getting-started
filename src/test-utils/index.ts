export * from "./axe-helper";
export * from "./Provider";
