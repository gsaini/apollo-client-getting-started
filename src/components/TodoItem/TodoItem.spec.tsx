import { fireEvent } from '@testing-library/react';
import React from 'react';
import { render } from '../../test-utils';
import { TodoItem } from './TodoItem';

describe('Todos components', () => {
    const todo = { id: "1", title: "lorem ipsum", completed: false };

    it('should render snapshot', async () => {
        const { container } = await render(<TodoItem todo={todo} onChange={jest.fn()} />);

        expect(container).toMatchSnapshot();
    });

    it('should toggle status', async () => {
        const onChangeMock = jest.fn();
        const { getByLabelText } = await render(<TodoItem todo={todo} onChange={onChangeMock} />);

        fireEvent.click(getByLabelText(todo.title));

        expect(onChangeMock).toHaveBeenCalledWith({ id: todo.id, completed: true });
    });
});