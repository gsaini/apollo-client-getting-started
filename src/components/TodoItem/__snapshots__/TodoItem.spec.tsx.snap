// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos components should render snapshot 1`] = `
<div>
  <div
    class="Todo__item"
  >
    <label
      for="1-checkbox"
    >
      lorem ipsum
    </label>
    <input
      id="1-checkbox"
      type="checkbox"
    />
  </div>
</div>
`;
